### PR TITLE
docs: document HTTP generic matching

### PR DESCRIPTION
**Pull request by snowp**
Signed-off-by: Snow Pettersen <snowp@lyft.com>

Risk Level: n/a
Testing: n/a
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto
Comment (htuch) [inline api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto:18]: Maybe "ignored for all data callbacks" rather than "ignored going".

File: api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto
Comment (htuch) [inline api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto:20]: This is the first time the concept of "data callback" comes up. Is there a way to explain it more clearly?

What happens when the matcher is used in a non-HTTP filter context? E.g. in some new router scheme. I guess HTTP filter callbacks still apply there, but what about TAP, etc.?

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:839]: Nit: s/a/an/ here and below before ExtensionWithMatcher.

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:8]: Nit: "makes heave use". I'd also maybe say "for performance better than linear list matching as seen in Envoy's HTTP routing" or the like to clearly provide the bogeyman that is at stake.

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:14]: @phlax should this be a verified code example? Can we generally just enable this for all new code fragments?

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:38]: Nit: prefer some RST formatting for filer proper names, e.g. `*SetResponseCode*`.

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:112]: Surprises are fun :) Is there a way to call out some concrete examples of where this happens?

**Review by htuch — CHANGES_REQUESTED** [review]
Super excited to see this land as a user facing feature :) I've left a lot of nits, but the docs looks great.
/wait

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (phlax) [reply docs/root/intro/arch_overview/matching/matching_api.rst:14]: the code that i have added verification for so far were complete configurations which are then added to the rst with `literalinclude`

we can do this here and validate just the yaml, or we could wrap this example in a complete config and just highlight the important parts

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (phlax) [reply docs/root/intro/arch_overview/matching/matching_api.rst:38]: should this be a link to the relevant filter?

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (snowp) [reply docs/root/intro/arch_overview/matching/matching_api.rst:112]: Sure I'll dig through and find an explicit example. 

Down the line I would like to use https://github.com/envoyproxy/envoy/issues/14727 to make it possible for filters to codify these to cause config rejection instead of these surprises, but this is what we've got for now

File: api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto
Comment (snowp) [reply api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto:20]: In my mind this would only be used in the context of HTTP filters (hence the name `SkipFilter`). Right now we don't have a mechanism to allowlist which actions can be used in various contexts, but I imagine we'll want this or some kind of template trickery to create unique sets of actions for different contexts. 

I've removed the use of "data callback", hopefully this will be easier to understand now

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (snowp) [reply docs/root/intro/arch_overview/matching/matching_api.rst:14]: Do I get validation for free just by following the literalinclude pattern used elsewhere? Or do I need to do anything else?

It's easy enough to include enough config to make it a valid config, so if we can easily validate the config by doing that then I think it's worth it

Comment (phlax) [reply docs/root/intro/arch_overview/matching/matching_api.rst:14]: its kinda the other way round in a sense

if you separate the config it will automatically get validated unless its specifically excluded

Comment (phlax) [reply docs/root/intro/arch_overview/matching/matching_api.rst:14]: ...at least for the envoy config - not sure about any yaml validation

Comment (phlax) [reply docs/root/intro/arch_overview/matching/matching_api.rst:14]: > Do I get validation for free just by following the literalinclude pattern used elsewhere? Or do I need to do anything else?

so, yep

**Comment by snowp** [comment]
This should be ready for another pass @htuch

File: api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto
Comment (htuch) [inline api/envoy/extensions/filters/common/matcher/action/v3/skip_action.proto:15]: Nit: RST markup for function literals.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:849]:
```suggestion
    // with the desired HTTP filter. This works for both the default filter configuration as well as for filters provided via the API.
```

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:22]: RST link to ExtensionMatcher def

File: docs/root/intro/arch_overview/matching/_include/simple.yaml
Comment (htuch) [inline docs/root/intro/arch_overview/matching/_include/simple.yaml:42]: It's a little confusing seeing both action name skip and the key here being `skip_filter`. Could the key (which is header value) be "some_value_to_match_on" or something like that?

File: docs/root/intro/arch_overview/matching/_include/complicated.yaml
Comment (htuch) [inline docs/root/intro/arch_overview/matching/_include/complicated.yaml:64]: A non-actionable comment that maybe we can just discuss thoughts on; the verbosity here is so high that I think we have two issues from a practical perspective:
1. Users who hand-write config are probably going to never want to use complex matchers. I feel the only way to use this reasonably is with some frontend language or API.
2. The config size is going to blow out at scale when we have type URL appearing everywhere. We've seen this also when looking at the impact of `xdstp://` URLs. It would be ironic if a solution aimed at scalability has a high cost due to config size explosion. I think for this problem we can maybe just add in some fixed matchers over time that take the place of common operations. This can be future work, but I think at scale it might be needed (e.g. a route table with 100k entries).

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:56]: Not sure about how to parse "By causing the filter to once the .."

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/matching/matching_api.rst:54]:
```suggestion
get partially applied until the match result is ready. This might result in surprising behavior. An example of this would be to have a matching
```

**Review by htuch — CHANGES_REQUESTED** [review]
@snowp thanks, generally looks great, just a few comments.
/wait

File: docs/root/intro/arch_overview/matching/_include/complicated.yaml
Comment (snowp) [reply docs/root/intro/arch_overview/matching/_include/complicated.yaml:64]: For 1 I agree that people are probably never going to write complicated trees by hand: even for existing APIs most the advanced use cases are probably written at least using protobuf intellisense/autocomplete. Do you think this example ends up being too complicated to be useful?

I think this is a valid concern, though the config size explosion mostly matters at config update, not stream processing. At some point it can become an issue (CPU usage, delayed config application) but I don't think it would directly affect data plane performance. I wonder if it makes sense to reuse the pattern of static SDS secrets where we could define named extensions somewhere and then simply reference them by name in config instead of repeating the whole configuration. That would avoid the explosion in cases where the exact same matcher/input is used in a lot of different branches

File: docs/root/intro/arch_overview/matching/_include/simple.yaml
Comment (phlax) [inline docs/root/intro/arch_overview/matching/_include/simple.yaml:67]: i think this should have a new line

File: docs/root/intro/arch_overview/matching/_include/complicated.yaml
Comment (phlax) [inline docs/root/intro/arch_overview/matching/_include/complicated.yaml:90]: also new line here

**Comment by phlax** [comment]
latest render is here https://storage.googleapis.com/envoy-pr/1419c99/docs/intro/arch_overview/matching/matching_api.html

not sure if its a tabs and spaces issue,, but some of the indents seem larger than others, which i think makes the config seem even more complicated in the complicated.yaml

**Comment by phlax** [comment]
...actually its more in the simple.yaml i think

**Comment by snowp** [comment]
@phlax good eye, thanks! Maybe a format linter for the YAML files would be a nice addition in the future

**Comment by phlax** [comment]
> Maybe a format linter for the YAML files would be a nice addition in the future

ive added a ticket re general file linting

ill look into yaml linting - i think it would be good to enforce least indent rules

File: docs/root/intro/arch_overview/matching/_include/complicated.yaml
Comment (htuch) [reply docs/root/intro/arch_overview/matching/_include/complicated.yaml]: Not arguing against including the example, just sharing some food for thought. I agree the cost is only at config update, that's where we're seeing real CPU costs at O(10k) cluster or route configs. 

We've been thinking even about templating to make xdstp:// names work better, I'll keep this use case in mind as well when thinking through that.

**Review by htuch — COMMENTED** [review]
LGTM, will defer to @mattklein123 @phlax for their comments.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:889]: Can you wire up the new `[#extension-category]` stuff here? See other recent examples. I'm surprised this wasn't already done. cc @phlax

File: docs/root/intro/arch_overview/matching/_include/complicated.yaml
Comment (mattklein123) [inline docs/root/intro/arch_overview/matching/_include/complicated.yaml:43]: This is really hard to read. Can you add more inline comments?

File: docs/root/intro/arch_overview/arch_overview.rst
Comment (mattklein123) [inline docs/root/intro/arch_overview/arch_overview.rst:10]: nit: I might consider moving this inside advanced? or at least lower in the list?

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (mattklein123) [inline docs/root/intro/arch_overview/matching/matching_api.rst:49]: Is it worth putting an attention block in here with a link to the open issue tracking match schema checking or similar?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Very cool! LGTM with small comments.

/wait

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:889]: I see this annotation is applied one layer higher. I would potentially move it here but up to you.

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (phlax) [inline docs/root/intro/arch_overview/matching/matching_api.rst:10]: s/performance better/better performance/

File: docs/root/intro/arch_overview/matching/matching_api.rst
Comment (phlax) [inline docs/root/intro/arch_overview/matching/matching_api.rst:66]: s/beahviour/behaviour/

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (phlax) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:889]: im not following i dont see the extension-category

**Comment by phlax** [comment]
current pages rendered:

https://storage.googleapis.com/envoy-pr/07f8b6d/docs/api-v3/extensions/filters/common/matcher/action/v3/skip_action.proto.html

https://storage.googleapis.com/envoy-pr/07f8b6d/docs/api-v3/extensions/common/matching/v3/extension_matcher.proto.html

https://storage.googleapis.com/envoy-pr/07f8b6d/docs/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto.html

**Comment by phlax** [comment]
not sure exactly where the extension/categories should be but i think they are missing

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (snowp) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:889]: There's `  // [#extension-category: envoy.filters.http]` on L304 on the HttpFilter field declaration

**Comment by phlax** [comment]
@snowp re-reading i think i understand - you can wrap any other extension using a matcher so no need to list out here

i think i was a bit confused by where the extension category is currently pinned - i think it needs to be moved to the `typed_config` - but that is unrelated to this PR

**Review by mattklein123 — APPROVED** [review]
LGTM modulo @phlax other PR. Thanks!

**Review by phlax — DISMISSED** [review]
lgtm - apologies for hesitation

**Comment by snowp** [comment]
@htuch Merging this is blocked on your review so please ptal :)

**Review by htuch — DISMISSED** [review]
Sorry, I LGTMed earlier but didn't approve.

**Comment by htuch** [comment]
@snowp needs main merge but otherwise good to go.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
